File: structural/bridge/notification_system/typescript/notification_system.ts
// --- Implementation Interface ---
export interface MessageSender {
  /**
   * Defines the interface for implementation classes (message sending mechanisms).
   */
  sendMessage(subject: string, body: string): void;
}

// --- Concrete Implementations ---
export class EmailSender implements MessageSender {
  /**
   * Concrete implementation for sending messages via Email.
   */
  public sendMessage(subject: string, body: string): void {
    console.log(`--- Sending Email ---`);
    console.log(`Subject: ${subject}`);
    console.log(`Body: ${body}`);
    console.log(`---------------------`);
  }
}

export class SmsSender implements MessageSender {
  /**
   * Concrete implementation for sending messages via SMS.
   */
  public sendMessage(subject: string, body: string): void {
    // SMS usually doesn't have a subject, so we combine them
    console.log(`--- Sending SMS ---`);
    console.log(`To: [PhoneNumber]`); // Placeholder
    console.log(`Message: ${subject} - ${body}`);
    console.log(`-----------------`);
  }
}

export class PushNotificationSender implements MessageSender {
  /**
   * Concrete implementation for sending messages via Push Notification.
   */
  public sendMessage(subject: string, body: string): void {
    console.log(`--- Sending Push Notification ---`);
    console.log(`Title: ${subject}`);
    console.log(`Body: ${body}`);
    console.log(`-------------------------------`);
  }
}

// --- Abstraction ---
export abstract class Notification {
  /**
   * Defines the abstraction's interface and maintains a reference to an implementation object.
   */
  protected sender: MessageSender; // Bridge

  constructor(sender: MessageSender) {
    this.sender = sender;
  }

  public abstract send(message: string): void;
}

// --- Refined Abstractions ---
export class InfoNotification extends Notification {
  /**
   * A specific type of notification (refined abstraction).
   */
  public send(message: string): void {
    const subject = "Info";
    const body = `[INFO] ${message}`;
    console.log(`Preparing '${subject}' notification.`);
    this.sender.sendMessage(subject, body);
  }
}

export class WarningNotification extends Notification {
  /**
   * Another specific type of notification.
   */
  public send(message: string): void {
    const subject = "Warning";
    const body = `[WARNING] ${message}`;
    console.log(`Preparing '${subject}' notification.`);
    this.sender.sendMessage(subject, body);
  }
}

export class UrgentNotification extends Notification {
  /**
   * An urgent notification type.
   */
  public send(message: string): void {
    const subject = "** URGENT **";
    const body = `[URGENT ACTION REQUIRED] ${message}`;
    console.log(`Preparing '${subject}' notification.`);
    this.sender.sendMessage(subject, body);
  }
}

File: structural/bridge/notification_system/python/notification_system.py
import abc

# --- Implementation Interface ---
class MessageSender(abc.ABC):
    """Defines the interface for implementation classes (message sending mechanisms)."""
    @abc.abstractmethod
    def send_message(self, subject: str, body: str):
        pass

# --- Concrete Implementations ---
class EmailSender(MessageSender):
    """Concrete implementation for sending messages via Email."""
    def send_message(self, subject: str, body: str):
        print(f"--- Sending Email ---")
        print(f"Subject: {subject}")
        print(f"Body: {body}")
        print(f"---------------------")

class SmsSender(MessageSender):
    """Concrete implementation for sending messages via SMS."""
    def send_message(self, subject: str, body: str):
        # SMS usually doesn't have a subject, so we combine them
        print(f"--- Sending SMS ---")
        print(f"To: [PhoneNumber]") # Placeholder
        print(f"Message: {subject} - {body}")
        print(f"-----------------")

class PushNotificationSender(MessageSender):
    """Concrete implementation for sending messages via Push Notification."""
    def send_message(self, subject: str, body: str):
        print(f"--- Sending Push Notification ---")
        print(f"Title: {subject}")
        print(f"Body: {body}")
        print(f"-------------------------------")

# --- Abstraction ---
class Notification(abc.ABC):
    """Defines the abstraction's interface and maintains a reference to an implementation object."""
    def __init__(self, sender: MessageSender):
        self._sender = sender # Bridge

    @abc.abstractmethod
    def send(self, message: str):
        pass

# --- Refined Abstractions ---
class InfoNotification(Notification):
    """A specific type of notification (refined abstraction)."""
    def send(self, message: str):
        subject = "Info"
        body = f"[INFO] {message}"
        print(f"Preparing '{subject}' notification.")
        self._sender.send_message(subject, body)

class WarningNotification(Notification):
    """Another specific type of notification."""
    def send(self, message: str):
        subject = "Warning"
        body = f"[WARNING] {message}"
        print(f"Preparing '{subject}' notification.")
        self._sender.send_message(subject, body)

class UrgentNotification(Notification):
    """An urgent notification type."""
    def send(self, message: str):
        subject = "** URGENT **"
        body = f"[URGENT ACTION REQUIRED] {message}"
        print(f"Preparing '{subject}' notification.")
        self._sender.send_message(subject, body)
